### PR TITLE
[Graph| tests, example, doc] Add GQA v2 support for implicit causal mask and example, doc update

### PR DESCRIPTION
**Pull request by xiang1guo**
Follow-up work based on #2931 

- [x] GQA v2 implicit causal mask test cases support
- [x] Add example for GQA v2 pattern.
- [x] Doc update.

### CODE REVIEW

File: examples/graph/gqa_v2.cpp [hunk 1]
```diff
@@ -0,0 +1,306 @@
+/*******************************************************************************
+* Copyright 2024-2025 Intel Corporation
```
Comment (TaoLv) [inline examples/graph/gqa_v2.cpp]: I assume that this new example reflects how GQA is integrated in PyTorch. @LuFinch could you please take a look at it?
If so, I would suggest to replace the original GQA example with it to avoid confusions.

Comment (xiang1guo) [reply examples/graph/gqa_v2.cpp]: I merged the 2 examples now, and also merged the doc change in this PR, @LuFinch please help to check, thanks~

File: examples/graph/gqa.cpp [hunk 2]
```diff
@@ -110,69 +110,46 @@ void bench_gqa(engine::kind ekind, logical_tensor::data_type dt,
...
 
     auto masked_score
-            = logical_tensor(id++, dt, score_sz, layout_type::strided);
+            = logical_tensor(id++, dt_inter, score_sz, layout_type::strided);
     auto mask_add = op(id++, op::kind::Add, "mask_add");
-    mask_add.add_inputs({scaled_score, mask_reshape});
+    mask_add.add_inputs({scaled_score, mask});
     mask_add.add_outputs({masked_score});
 
     // attention_probs = softmax(masked_score)
```
Comment (TaoLv) [inline examples/graph/gqa.cpp:155]: Please set attribute `inf_as_zero` for softmax.

Comment (xiang1guo) [reply examples/graph/gqa.cpp:155]: Added, thanks for reminding~

File: examples/graph/gqa_v2.cpp (hunk 1 above)
Comment (LuFinch) [reply examples/graph/gqa_v2.cpp]: Check the doc and gqa.example. Pytorch SDPA XPU integrated GQAv2 now. Overall LGTM.

**Comment by xiang1guo** [comment]
make test
set test_scope=NIGHTLY
disable benchdnn_all
enable benchdnn_graph

**Comment by xiang1guo** [comment]
make test
set test_scope=NIGHTLY
disable benchdnn_all
enable benchdnn_graph

File: doc/graph/fusion_patterns/gqa.md [hunk 3]
```diff
@@ -30,14 +30,20 @@ The notations used in the document:
 
 Similar to how SDPA is supported, the GQA pattern is also defined as a
 directional acyclic graph (DAG) using oneDNN Graph API. oneDNN extends the
-[SDPA pattern](@ref dev_guide_graph_sdpa) to support floating-point (f32, bf16,
-and f16) GQA as follows. The blue nodes are required when defining a GQA pattern
-while the brown nodes are optional.
+[SDPA pattern](@ref dev_guide_graph_sdpa) to support 2 types of floating-point
+(f32, bf16, and f16) GQA patterns . The blue nodes are required when defining a
```
Comment (ElaineBao) [inline doc/graph/fusion_patterns/gqa.md]: Do you think it will be better to merge the two images into one, while marking StaticReshape as brown nodes? You may add descriptions to tell StaticReshape is used to reshape 4D to 5D or 5D to 4D if inputs and outputs are 4D.

**Review by TaoLv — APPROVED** [review]

File: doc/graph/fusion_patterns/gqa.md [hunk 4]
```diff
@@ -30,14 +30,24 @@ The notations used in the document:
 
 Similar to how SDPA is supported, the GQA pattern is also defined as a
 directional acyclic graph (DAG) using oneDNN Graph API. oneDNN extends the
-[SDPA pattern](@ref dev_guide_graph_sdpa) to support floating-point (f32, bf16,
-and f16) GQA as follows. The blue nodes are required when defining a GQA pattern
-while the brown nodes are optional.
+[SDPA pattern](@ref dev_guide_graph_sdpa) to support two types of floating-point
+(f32, bf16, and f16) GQA patterns . The blue nodes are required when defining a
```
Comment (ElaineBao) [inline doc/graph/fusion_patterns/gqa.md]:
```suggestion
(f32, bf16, and f16) GQA patterns. The blue nodes are required when defining a
```

File: doc/graph/fusion_patterns/gqa.md [hunk 5]
```diff
@@ -30,14 +30,24 @@ The notations used in the document:
...
+GQA pattern while the brown nodes are optional. The key difference between the
+two types of GQA patterns lies in whether the input and output tensors have a 4D
+or 5D shape. The optional StaticReshape operation is used to convert the tensor
+between 4D and 5D shape formats, depending on whether the input and output tensor
+is in 4D shape.
```
Comment (ElaineBao) [inline doc/graph/fusion_patterns/gqa.md]:
```suggestion
two types of GQA patterns lies in whether the input and output tensors have 4D
or 5D shapes. The optional StaticReshape operations are used to convert the tensors
between 4D and 5D shape formats, depending on whether the input and output tensors
are in 4D shapes.
```

File: doc/graph/fusion_patterns/gqa.md (hunk 4 above)
Comment (xiang1guo) [reply doc/graph/fusion_patterns/gqa.md]: Updated! Thanks!

File: doc/graph/fusion_patterns/gqa.md (hunk 5 above)
Comment (xiang1guo) [reply doc/graph/fusion_patterns/gqa.md]: updated! Thanks again~

**Review by ElaineBao — APPROVED** [review]

**Comment by xiang1guo** [comment]
make test
set test_scope=NIGHTLY
disable benchdnn_all
enable benchdnn_graph

**Review by dzarukin — APPROVED** [review]

File: doc/graph/fusion_patterns/gqa.md [hunk 6]
```diff
@@ -56,6 +66,19 @@ pattern:
    similarly. Besides that, they have the same definition as described in the
    typical SDPA pattern.
 
+### GQA Pattern with 5D input and output
+
+To simplify process and avoid unnecessary reshapes, oneDNN also supports native
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md]:
```suggestion
To simplify the process and avoid unnecessary reshaping, the native 5D GQA pattern supported by oneDNN can be used.
```

File: doc/graph/fusion_patterns/gqa.md [hunk 7]
```diff
@@ -56,6 +66,19 @@ pattern:
...
+5D GQA pattern. In this approach, the input Query, Key, and Value tensors are
+already provided in grouped format.
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md]:
```suggestion
provided in a grouped format.
```

File: doc/graph/fusion_patterns/gqa.md [hunk 8]
```diff
@@ -56,6 +66,19 @@ pattern:
...
+
+1. The input Query has 5D shape: (N, H_kv, N_rep, S, D)
+2. The input Key/Value have 5D shape: (N, H_kv, 1, S, D)
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md]: The input paramater can contain either a Key or Value? IF it's an either/or situation, then singular verb should be used.
```suggestion
2. The input Key/Value has a 5D shape: (N, H_kv, 1, S, D)
```

File: doc/graph/fusion_patterns/gqa.md [hunk 9]
```diff
@@ -56,6 +66,19 @@ pattern:
    similarly. Besides that, they have the same definition as described in the
    typical SDPA pattern.
 
+### GQA Pattern with 5D input and output
+
+To simplify process and avoid unnecessary reshapes, oneDNN also supports native
+5D GQA pattern. In this approach, the input Query, Key, and Value tensors are
+already provided in grouped format.
+
+1. The input Query has 5D shape: (N, H_kv, N_rep, S, D)
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md]:
```suggestion
1. The input Query has a 5D shape: (N, H_kv, N_rep, S, D)
```

File: doc/graph/fusion_patterns/gqa.md [hunk 10]
```diff
@@ -56,6 +66,19 @@ pattern:
...
+2. The input Key/Value have 5D shape: (N, H_kv, 1, S, D)
+3. The second MatMul calculates the dot products between the probabilities after
+   SoftMax and Value nodes and generates output with shape (N, H_kv, N_rep, S, D).
+4. The input scale factor and mask in the pattern also need to meet the
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md]:
```suggestion
4. The input scale factor and mask in the pattern must meet the
```

**Review by ranukund — COMMENTED** [review]
Few edits suggested, please incorporate as you see fit., thanks

**Comment by xiang1guo** [comment]
> Few edits suggested, please incorporate as you see fit., thanks

Thanks @ranukund ! Updated the doc according to your suggestions!

**Comment by xiang1guo** [comment]
make test
set test_scope=NIGHTLY
disable benchdnn_all
enable benchdnn_graph

File: doc/graph/fusion_patterns/gqa.md [hunk 11]
```diff
@@ -30,14 +30,24 @@ The notations used in the document:
...
 
 ![GQA pattern](images/gqa.png)
 
-Compared to a typical SDPA pattern, there are a few differences in the GQA
-pattern:
+### GQA Pattern with 4D input and output
+
+Due to the broadcasting semantics of MatMul, implementing GQA often requires
+additional tensor manipulation. Specifically, when working with 4D input tensors,
+where Query has shape (N, H_q, S, D) and Key/Value have shape (N, H_kv, S, D),
```
Comment (ranukund) [inline doc/graph/fusion_patterns/gqa.md]:
```suggestion
where Query has shape (N, H_q, S, D) and Key/Value has shape (N, H_kv, S, D),
```

**Review by ranukund — APPROVED** [review]
A minor edit, please incorporate, thanks!

File: doc/graph/fusion_patterns/gqa.md (hunk 11 above)
Comment (xiang1guo) [reply doc/graph/fusion_patterns/gqa.md]: Thanks @ranukund ! Updated~

**Comment by xiang1guo** [comment]
make test
set test_scope=NIGHTLY
disable benchdnn_all
enable benchdnn_graph

**Comment by xiang1guo** [comment]
make test
set test_scope=NIGHTLY
disable benchdnn_all
enable benchdnn_graph
